src/CanonicalizeGPUVars.cpp

```diff
@@ -29,51 +29,42 @@ string get_block_name(int index) {
 }
 
 class CountGPUBlocksThreads : public IRVisitor {
-    string prefix;  // Producer name + stage
-
     using IRVisitor::visit;
 
     void visit(const For *op) override {
-        if (starts_with(op->name, prefix)) {
-            if (op->for_type == ForType::GPUBlock) {
-                nblocks++;
-            } else if (op->for_type == ForType::GPUThread) {
-                nthreads++;
-            } else if (op->for_type == ForType::GPULane) {
-                nlanes++;
-            }
-        }
+        // Figure out how much to increment each counter by based on the loop
+        // type.
+        int db = op->for_type == ForType::GPUBlock;
+        int dl = op->for_type == ForType::GPULane;
+        int dt = op->for_type == ForType::GPUThread;
+
+        // The threads counter includes lanes loops
+        dt += dl;
+
+        // Increment counters
+        nb += db;
+        nl += dl;
+        nt += dt;
+
+        // Update the high water mark.
+        nblocks = std::max(nb, nblocks);
+        nthreads = std::max(nt, nthreads);
+        nlanes = std::max(nl, nlanes);
+
+        // Visit the body
         IRVisitor::visit(op);
-    }
 
-    void visit(const IfThenElse *op) override {
-        op->condition.accept(this);
-
-        int old_nblocks = nblocks;
-        int old_nthreads = nthreads;
-        int old_nlanes = nlanes;
-        op->then_case.accept(this);
-
-        if (op->else_case.defined()) {
-            int then_nblocks = nblocks;
-            int then_nthreads = nthreads;
-            int then_nlanes = nlanes;
-            nblocks = old_nblocks;
-            nthreads = old_nthreads;
-            nlanes = old_nlanes;
-            op->else_case.accept(this);
-            nblocks = std::max(then_nblocks, nblocks);
-            nthreads = std::max(then_nthreads, nthreads);
-            nlanes = std::max(then_nlanes, nlanes);
-        }
+        // Decrement counters
+        nb -= db;
+        nl -= dl;
+        nt -= dt;
     }
 
+    int nb = 0, nt = 0, nl = 0;
```

Comment:
I think this needs more documentation. How are these counters different from `nblocks`, `nthreads`, and `nlanes`? I can't tell from the existing code / comments in the `For` visitor.